src/Compilers/CSharp/Portable/Lowering/SyntheticBoundNodeFactory.cs

```diff
@@ -747,8 +747,22 @@ public BoundExpression Sequence(params BoundExpression[] parts)
         public BoundExpression Sequence(ImmutableArray<LocalSymbol> locals, params BoundExpression[] parts)
         {
             var builder = ArrayBuilder<BoundExpression>.GetInstance();
-            for (int i = 0; i < parts.Length - 1; i++) builder.Add(parts[i]);
+            for (int i = 0; i < parts.Length - 1; i++)
+            {
+                var part = parts[i];
+                if (LocalRewriter.ReadIsSideeffecting(part))
```

Comment:
That would be after lambdas/closures and everything else. There is a benefit in trimming complexity of the bound trees early since optimizations tend to have cascading effect. Also some cases are just easier to handle here before trees are transformed.

Codegen does have optimizations similar to this. Mostly to pick up remaining redundancies or if something is introduced while lowering.